File: api/src/main/java/com/ruby/rubia_server/core/service/UserService.java
package com.ruby.rubia_server.core.service;

import com.ruby.rubia_server.core.dto.CreateUserDTO;
import com.ruby.rubia_server.core.dto.UpdateUserDTO;
import com.ruby.rubia_server.core.dto.UserDTO;
import com.ruby.rubia_server.core.dto.UserLoginDTO;
import com.ruby.rubia_server.core.entity.Company;
import com.ruby.rubia_server.core.entity.Department;
import com.ruby.rubia_server.core.entity.User;
import com.ruby.rubia_server.core.enums.UserRole;
import com.ruby.rubia_server.core.repository.CompanyRepository;
import com.ruby.rubia_server.core.repository.DepartmentRepository;
import com.ruby.rubia_server.core.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Slf4j
@Transactional
public class UserService {
    
    private final UserRepository userRepository;
    private final DepartmentRepository departmentRepository;
    private final CompanyRepository companyRepository;
    private final BCryptPasswordEncoder passwordEncoder = new BCryptPasswordEncoder();
    
    public UserDTO create(CreateUserDTO createDTO) {
        log.info("Creating user with email: {}", createDTO.getEmail());
        
        if (userRepository.existsByEmailAndCompanyId(createDTO.getEmail(), createDTO.getCompanyId())) {
            throw new IllegalArgumentException("Usuário com email '" + createDTO.getEmail() + "' já existe nesta empresa");
        }
        
        Company company = companyRepository.findById(createDTO.getCompanyId())
                .orElseThrow(() -> new IllegalArgumentException("Empresa não encontrada"));
        
        Department department = null;
        if (createDTO.getDepartmentId() != null) {
            department = departmentRepository.findById(createDTO.getDepartmentId())
                    .orElseThrow(() -> new IllegalArgumentException("Departamento não encontrado"));
        }
        
        User user = User.builder()
                .name(createDTO.getName())
                .email(createDTO.getEmail())
                .passwordHash(passwordEncoder.encode(createDTO.getPassword()))
                .company(company)
                .department(department)
                .role(createDTO.getRole())
                .avatarUrl(createDTO.getAvatarUrl())
                .birthDate(createDTO.getBirthDate())
                .weight(createDTO.getWeight())
                .height(createDTO.getHeight())
                .address(createDTO.getAddress())
                .isOnline(false)
                .build();
        
        User saved = userRepository.save(user);
        log.info("User created successfully with id: {}", saved.getId());
        
        return toDTO(saved);
    }
    
    @Transactional(readOnly = true)
    public UserDTO findById(UUID id) {
        log.debug("Finding user by id: {}", id);
        
        User user = userRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Usuário não encontrado"));
        
        return toDTO(user);
    }
    
    @Transactional(readOnly = true)
    public UserDTO findByEmailAndCompany(String email, UUID companyId) {
        log.debug("Finding user by email: {} for company: {}", email, companyId);
        
        User user = userRepository.findByEmailAndCompanyId(email, companyId)
                .orElseThrow(() -> new IllegalArgumentException("Usuário não encontrado nesta empresa"));
        
        return toDTO(user);
    }
    
    @Transactional(readOnly = true)
    public List<UserDTO> findAllByCompany(UUID companyId) {
        log.debug("Finding all users for company: {}", companyId);
        
        return userRepository.findByCompanyId(companyId)
                .stream()
                .map(this::toDTO)
                .toList();
    }
    
    @Transactional(readOnly = true)
    public List<UserDTO> findByDepartmentAndCompany(UUID departmentId, UUID companyId) {
        log.debug("Finding users by department: {} for company: {}", departmentId, companyId);
        
        return userRepository.findByDepartmentIdAndCompanyId(departmentId, companyId)
                .stream()
                .map(this::toDTO)
                .toList();
    }
    
    @Transactional(readOnly = true)
    public List<UserDTO> findAvailableAgentsByCompany(UUID companyId) {
        log.debug("Finding available agents for company: {}", companyId);
        
        return userRepository.findAvailableAgentsByCompany(companyId)
                .stream()
                .map(this::toDTO)
                .toList();
    }
    
    @Transactional(readOnly = true)
    public List<UserDTO> findAvailableAgentsByDepartmentAndCompany(UUID departmentId, UUID companyId) {
        log.debug("Finding available agents by department: {} for company: {}", departmentId, companyId);
        
        return userRepository.findAvailableAgentsByDepartmentAndCompany(departmentId, companyId)
                .stream()
                .map(this::toDTO)
                .toList();
    }
    
    public UserDTO update(UUID id, UpdateUserDTO updateDTO) {
        log.info("Updating user with id: {}", id);
        
        User user = userRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Usuário não encontrado"));
        
        if (updateDTO.getName() != null) {
            user.setName(updateDTO.getName());
        }
        
        if (updateDTO.getEmail() != null) {
            if (!updateDTO.getEmail().equals(user.getEmail()) && 
                userRepository.existsByEmailAndCompanyId(updateDTO.getEmail(), user.getCompany().getId())) {
                throw new IllegalArgumentException("Usuário com email '" + updateDTO.getEmail() + "' já existe nesta empresa");
            }
            user.setEmail(updateDTO.getEmail());
        }
        
        if (updateDTO.getPassword() != null) {
            user.setPasswordHash(passwordEncoder.encode(updateDTO.getPassword()));
        }
        
        if (updateDTO.getDepartmentId() != null) {
            Department department = departmentRepository.findById(updateDTO.getDepartmentId())
                    .orElseThrow(() -> new IllegalArgumentException("Departamento não encontrado"));
            user.setDepartment(department);
        }
        
        if (updateDTO.getRole() != null) {
            user.setRole(updateDTO.getRole());
        }
        
        if (updateDTO.getAvatarUrl() != null) {
            user.setAvatarUrl(updateDTO.getAvatarUrl());
        }
        
        if (updateDTO.getBirthDate() != null) {
            user.setBirthDate(updateDTO.getBirthDate());
        }
        
        if (updateDTO.getWeight() != null) {
            user.setWeight(updateDTO.getWeight());
        }
        
        if (updateDTO.getHeight() != null) {
            user.setHeight(updateDTO.getHeight());
        }
        
        if (updateDTO.getAddress() != null) {
            user.setAddress(updateDTO.getAddress());
        }
        
        User updated = userRepository.save(user);
        log.info("User updated successfully");
        
        return toDTO(updated);
    }
    
    public UserDTO updateOnlineStatus(UUID id, boolean isOnline) {
        log.info("Updating online status for user: {} to {}", id, isOnline);
        
        User user = userRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Usuário não encontrado"));
        
        user.setIsOnline(isOnline);
        if (!isOnline) {
            user.setLastSeen(LocalDateTime.now());
        }
        
        User updated = userRepository.save(user);
        log.info("User online status updated successfully");
        
        return toDTO(updated);
    }
    
    public UserDTO assignToDepartment(UUID userId, UUID departmentId) {
        log.info("Assigning user {} to department {}", userId, departmentId);
        
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new IllegalArgumentException("Usuário não encontrado"));
        
        Department department = departmentRepository.findById(departmentId)
                .orElseThrow(() -> new IllegalArgumentException("Departamento não encontrado"));
        
        user.setDepartment(department);
        User updated = userRepository.save(user);
        
        log.info("User assigned to department successfully");
        return toDTO(updated);
    }
    
    @Transactional(readOnly = true)
    public boolean validateLoginByCompany(UserLoginDTO loginDTO, UUID companyId) {
        log.debug("Validating login for email: {} in company: {}", loginDTO.getEmail(), companyId);
        
        User user = userRepository.findByEmailAndCompanyId(loginDTO.getEmail(), companyId)
                .orElse(null);
        
        if (user == null) {
            return false;
        }
        
        return passwordEncoder.matches(loginDTO.getPassword(), user.getPasswordHash());
    }
    
    public void delete(UUID id) {
        log.info("Deleting user with id: {}", id);
        
        if (!userRepository.existsById(id)) {
            throw new IllegalArgumentException("Usuário não encontrado");
        }
        
        userRepository.deleteById(id);
        log.info("User deleted successfully");
    }
    
    private UserDTO toDTO(User user) {
        return UserDTO.builder()
                .id(user.getId())
                .name(user.getName())
                .email(user.getEmail())
                .companyId(user.getCompany() != null ? user.getCompany().getId() : null)
                .departmentId(user.getDepartment() != null ? user.getDepartment().getId() : null)
                .departmentName(user.getDepartment() != null ? user.getDepartment().getName() : null)
                .role(user.getRole())
                .avatarUrl(user.getAvatarUrl())
                .isOnline(user.getIsOnline())
                .lastSeen(user.getLastSeen())
                .birthDate(user.getBirthDate())
                .weight(user.getWeight())
                .height(user.getHeight())
                .address(user.getAddress())
                .createdAt(user.getCreatedAt())
                .updatedAt(user.getUpdatedAt())
                .build();
    }
}

File: api-dotnet/Rubia.Server/Services/UserService.cs
using Microsoft.EntityFrameworkCore;
using Rubia.Server.Data;
using Rubia.Server.DTOs;
using Rubia.Server.Entities;
using Rubia.Server.Enums;
using Rubia.Server.Services.Interfaces;

namespace Rubia.Server.Services;

public class UserService : IUserService
{
    private readonly RubiaDbContext _context;
    private readonly ILogger<UserService> _logger;

    public UserService(RubiaDbContext context, ILogger<UserService> logger)
    {
        _context = context;
        _logger = logger;
    }

    public async Task<UserDto> CreateAsync(CreateUserDto createDto)
    {
        _logger.LogInformation("Creating user with email: {Email}", createDto.Email);

        if (await _context.Users.AnyAsync(u => u.Email == createDto.Email && u.CompanyId == createDto.CompanyId))
        {
            throw new ArgumentException($"Usuário com email '{createDto.Email}' já existe nesta empresa");
        }

        var company = await _context.Companies.FirstOrDefaultAsync(c => c.Id == createDto.CompanyId);
        if (company == null)
        {
            throw new ArgumentException("Empresa não encontrada");
        }

        var department = await _context.Departments.FirstOrDefaultAsync(d => d.Id == createDto.DepartmentId);
        if (department == null)
        {
            throw new ArgumentException("Departamento não encontrado");
        }

        var user = new User
        {
            Name = createDto.Name,
            Email = createDto.Email,
            PasswordHash = BCrypt.Net.BCrypt.HashPassword(createDto.Password),
            CompanyId = createDto.CompanyId,
            DepartmentId = createDto.DepartmentId,
            Role = createDto.Role,
            AvatarUrl = createDto.AvatarUrl,
            BirthDate = createDto.BirthDate,
            Weight = createDto.Weight,
            Height = createDto.Height,
            Address = createDto.Address,
            IsOnline = false
        };

        _context.Users.Add(user);
        await _context.SaveChangesAsync();

        // Load related entities for DTO conversion
        await _context.Entry(user)
            .Reference(u => u.Department)
            .LoadAsync();

        _logger.LogInformation("User created successfully with id: {Id}", user.Id);
        return ToDto(user);
    }

    public async Task<UserDto> FindByIdAsync(Guid id)
    {
        _logger.LogDebug("Finding user by id: {Id}", id);

        var user = await _context.Users
            .Include(u => u.Department)
            .FirstOrDefaultAsync(u => u.Id == id);

        if (user == null)
        {
            throw new ArgumentException("Usuário não encontrado");
        }

        return ToDto(user);
    }

    public async Task<UserDto> FindByEmailAndCompanyAsync(string email, Guid companyId)
    {
        _logger.LogDebug("Finding user by email: {Email} for company: {CompanyId}", email, companyId);

        var user = await _context.Users
            .Include(u => u.Department)
            .FirstOrDefaultAsync(u => u.Email == email && u.CompanyId == companyId);

        if (user == null)
        {
            throw new ArgumentException("Usuário não encontrado nesta empresa");
        }

        return ToDto(user);
    }

    public async Task<List<UserDto>> FindAllByCompanyAsync(Guid companyId)
    {
        _logger.LogDebug("Finding all users for company: {CompanyId}", companyId);

        var users = await _context.Users
            .Include(u => u.Department)
            .Where(u => u.CompanyId == companyId)
            .OrderBy(u => u.Name)
            .ToListAsync();

        return users.Select(ToDto).ToList();
    }

    public async Task<List<UserDto>> FindByDepartmentAndCompanyAsync(Guid departmentId, Guid companyId)
    {
        _logger.LogDebug("Finding users by department: {DepartmentId} for company: {CompanyId}", 
            departmentId, companyId);

        var users = await _context.Users
            .Include(u => u.Department)
            .Where(u => u.DepartmentId == departmentId && u.CompanyId == companyId)
            .OrderBy(u => u.Name)
            .ToListAsync();

        return users.Select(ToDto).ToList();
    }

    public async Task<List<UserDto>> FindAvailableAgentsByCompanyAsync(Guid companyId)
    {
        _logger.LogDebug("Finding available agents for company: {CompanyId}", companyId);

        var users = await _context.Users
            .Include(u => u.Department)
            .Where(u => u.CompanyId == companyId && 
                       (u.Role == UserRole.Agent || u.Role == UserRole.Supervisor || u.Role == UserRole.Admin))
            .OrderBy(u => u.Name)
            .ToListAsync();

        return users.Select(ToDto).ToList();
    }

    public async Task<List<UserDto>> FindAvailableAgentsByDepartmentAndCompanyAsync(Guid departmentId, Guid companyId)
    {
        _logger.LogDebug("Finding available agents by department: {DepartmentId} for company: {CompanyId}", 
            departmentId, companyId);

        var users = await _context.Users
            .Include(u => u.Department)
            .Where(u => u.DepartmentId == departmentId && 
                       u.CompanyId == companyId && 
                       (u.Role == UserRole.Agent || u.Role == UserRole.Supervisor || u.Role == UserRole.Admin))
            .OrderBy(u => u.Name)
            .ToListAsync();

        return users.Select(ToDto).ToList();
    }

    public async Task<UserDto> UpdateAsync(Guid id, UpdateUserDto updateDto)
    {
        _logger.LogInformation("Updating user with id: {Id}", id);

        var user = await _context.Users
            .Include(u => u.Department)
            .FirstOrDefaultAsync(u => u.Id == id);

        if (user == null)
        {
            throw new ArgumentException("Usuário não encontrado");
        }

        if (updateDto.Name != null) user.Name = updateDto.Name;

        if (updateDto.Email != null)
        {
            if (updateDto.Email != user.Email && 
                await _context.Users.AnyAsync(u => u.Email == updateDto.Email && u.CompanyId == user.CompanyId))
            {
                throw new ArgumentException($"Usuário com email '{updateDto.Email}' já existe nesta empresa");
            }
            user.Email = updateDto.Email;
        }

        if (updateDto.Password != null)
        {
            user.PasswordHash = BCrypt.Net.BCrypt.HashPassword(updateDto.Password);
        }

        if (updateDto.DepartmentId.HasValue)
        {
            var department = await _context.Departments.FirstOrDefaultAsync(d => d.Id == updateDto.DepartmentId);
            if (department == null)
            {
                throw new ArgumentException("Departamento não encontrado");
            }
            user.DepartmentId = updateDto.DepartmentId.Value;
        }

        if (updateDto.Role.HasValue) user.Role = updateDto.Role.Value;
        if (updateDto.AvatarUrl != null) user.AvatarUrl = updateDto.AvatarUrl;
        if (updateDto.BirthDate.HasValue) user.BirthDate = updateDto.BirthDate;
        if (updateDto.Weight.HasValue) user.Weight = updateDto.Weight;
        if (updateDto.Height.HasValue) user.Height = updateDto.Height;
        if (updateDto.Address != null) user.Address = updateDto.Address;

        await _context.SaveChangesAsync();
        
        // Reload department if changed
        if (updateDto.DepartmentId.HasValue)
        {
            await _context.Entry(user)
                .Reference(u => u.Department)
                .LoadAsync();
        }

        _logger.LogInformation("User updated successfully");
        return ToDto(user);
    }

    public async Task<UserDto> UpdateOnlineStatusAsync(Guid id, bool isOnline)
    {
        _logger.LogInformation("Updating online status for user: {Id} to {IsOnline}", id, isOnline);

        var user = await _context.Users
            .Include(u => u.Department)
            .FirstOrDefaultAsync(u => u.Id == id);

        if (user == null)
        {
            throw new ArgumentException("Usuário não encontrado");
        }

        user.IsOnline = isOnline;
        if (!isOnline)
        {
            user.LastSeen = DateTime.UtcNow;
        }

        await _context.SaveChangesAsync();
        _logger.LogInformation("User online status updated successfully");

        return ToDto(user);
    }

    public async Task<UserDto> AssignToDepartmentAsync(Guid userId, Guid departmentId)
    {
        _logger.LogInformation("Assigning user {UserId} to department {DepartmentId}", userId, departmentId);

        var user = await _context.Users
            .Include(u => u.Department)
            .FirstOrDefaultAsync(u => u.Id == userId);

        if (user == null)
        {
            throw new ArgumentException("Usuário não encontrado");
        }

        var department = await _context.Departments.FirstOrDefaultAsync(d => d.Id == departmentId);
        if (department == null)
        {
            throw new ArgumentException("Departamento não encontrado");
        }

        user.DepartmentId = departmentId;
        await _context.SaveChangesAsync();

        // Reload department
        await _context.Entry(user)
            .Reference(u => u.Department)
            .LoadAsync();

        _logger.LogInformation("User assigned to department successfully");
        return ToDto(user);
    }

    public async Task<bool> ValidateLoginByCompanyAsync(UserLoginDto loginDto, Guid companyId)
    {
        _logger.LogDebug("Validating login for email: {Email} in company: {CompanyId}", loginDto.Email, companyId);

        var user = await _context.Users
            .FirstOrDefaultAsync(u => u.Email == loginDto.Email && u.CompanyId == companyId);

        if (user == null)
        {
            return false;
        }

        return BCrypt.Net.BCrypt.Verify(loginDto.Password, user.PasswordHash);
    }

    public async Task DeleteAsync(Guid id)
    {
        _logger.LogInformation("Deleting user with id: {Id}", id);

        var user = await _context.Users.FirstOrDefaultAsync(u => u.Id == id);
        if (user == null)
        {
            throw new ArgumentException("Usuário não encontrado");
        }

        _context.Users.Remove(user);
        await _context.SaveChangesAsync();

        _logger.LogInformation("User deleted successfully");
    }

    public async Task DeleteAllByCompanyAsync(Guid companyId)
    {
        _logger.LogInformation("Deleting all users for company: {CompanyId}", companyId);

        var users = await _context.Users.Where(u => u.CompanyId == companyId).ToListAsync();
        _context.Users.RemoveRange(users);
        await _context.SaveChangesAsync();

        _logger.LogInformation("Deleted {Count} users for company: {CompanyId}", users.Count, companyId);
    }

    private static UserDto ToDto(User user)
    {
        return new UserDto
        {
            Id = user.Id,
            Name = user.Name,
            Email = user.Email,
            CompanyId = user.CompanyId,
            DepartmentId = user.DepartmentId,
            DepartmentName = user.Department?.Name,
            Role = user.Role,
            AvatarUrl = user.AvatarUrl,
            IsOnline = user.IsOnline,
            LastSeen = user.LastSeen,
            BirthDate = user.BirthDate,
            Weight = user.Weight,
            Height = user.Height,
            Address = user.Address,
            CreatedAt = user.CreatedAt,
            UpdatedAt = user.UpdatedAt
        };
    }
}